man/hfst-regexp2fst.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.38.4.
.TH HFST-REGEXP2FST "1" "March 2011" "HFST" "User Commands"
.SH NAME
hfst-regexp2fst \- manual page for hfst-regexp2fst 0.1 (hfst 3.0)
.SH SYNOPSIS
.B hfst-regexp2fst
[\fIOPTIONS\fR...] [\fIINFILE\fR]
.SH DESCRIPTION
Compile string pairs and pair\-strings into transducer(s). The format for
regular expression for describing two-level automata is detailed in the
book "Finite-State Morphology" by Kenneth R. Beesley and Lauri Karttunen.
\" for reference see 
\" \BR regex_xerox (7)
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Do not print output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fIINFILE\fR
Read input transducer from \fIINFILE\fR
.TP
\fB\-o\fR, \fB\-\-output\fR=\fIOUTFILE\fR
Write output transducer to \fIOUTFILE\fR
.SS "String and format options:"
.TP
\fB\-f\fR, \fB\-\-format\fR=\fIFMT\fR
Write result in \fIFMT\fR format
.TP
\fB\-j\fR, \fB\-\-disjunct\fR
Disjunct instead of transforming each expression into a separate transducer
.TP
\fB\-l\fR, \fB\-\-line\fR
Input is line separated
.TP
\fB\-S\fR, \fB\-\-semicolon\fR
Input is semicolon separated
.TP
\fB\-e\fR, \fB\-\-epsilon\fR=\fIEPS\fR
Map \fIEPS\fR as epsilon label
.PP
If \fIOUTFILE\fR or \fIINFILE\fR is missing or \-, standard streams will be
used. \fIFMT\fR must be name of a format usable by libhfst, such as 
\fBopenfst\-tropical\fR, \fBsfst\fR, \fBfoma\fR or \fBolw\fR.
If \fIEPS\fR is not defined, the default representation of \fB0\fR is used
.SH EXAMPLES
.TP
echo "c:d a:o t:g" | hfst\-regexp2fst \fB\-l\fR
create cat:dog fst
.TP
echo "c:d a:o t:g ;" | hfst\-regexp2fst \fB\-S\fR
same as pairstring
.SS "Input format:"
The format for regular expressions supported by regexp2fst is a subset of
regular expression format in
.BR xfst
that has been implemented.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://sourceforge.net/tracker/?atid=1061990&group_id=224521&func=browse>
.PP
hfst\-regexp2fst home page:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//Regexp2Fst>
.br
General help using HFST software:
<https://kitwiki.csc.fi/twiki/bin/view/KitWiki//HfstHome>
.SH COPYRIGHT
Copyright \(co 2010 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH AUTHOR
Written by HFST team <hfst-bugs@helsinki.fi>.


NEWS
=================================
 NEWS for hfst-tools and library
=================================

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

Noteworthy changes in 3.0.2
---------------------------

* some fixes to OSX compatibility

Noteworthy changes in 3.0.1
---------------------------

* this is a bugfix release

* conversion to hfst-optimized-lookup format was fundamentally broken in the
  original release; this should provide a fix and considerable speedup

* compose-intersect memory use is much improved

Noteworthy changes in 3.0
-------------------------

* all new API in object oriented form

* brand new foma backend for unweighted transducer handling

* the FST libraries used for automata handling are now dynamically linked to
  HFST

    * this requires end users to install needed backends themselves

* hfst-lexc and hfst-xfst based on foma's compiler

  * xfst is now also fully supported

  * the obsolete command line clients for HFST-based compilation of lexc and
    xfst scripts are hidden in hfst-lexc2fst and hfst-xfst2fst

* preliminary support for unknown character classes

* optimized-lookup backend supported in library

  * dropped special tools for optimised handling since optimised lookup is
    included in library: hfst-optimised-lookup -> hfst-lookup; 
    hfst-lookup-optimize -> hfst-fst2fst

* hfst-proc tool to emulate lt-proc, cg-proc

* Slightly improved string and text formats to support weights uniformly and
  parsing tabs and spaces as automata arcs: see `wiki page for text and string
  formats
  <https://kitwiki.csc.fi/twiki/bin/view/KitWiki/HfstStringAndTextFormats>`

* better error handling using exceptions

* The argument handling order of binary tools changed: stdin is first if not
  specified

* The automata are now packed in HFST automaton container format (akin to AVI
  for videos); this supports arbitrary metadata and underlying FST formats

* the symbol handling is more implicit; no more symbol table arguments for tools
  or hfst-symbols command

* converted to new autotools: silent rules and color tests

* Hfst's internal automata format for fallback on some unsupported operations
  and for conversions

Noteworthy changes in 2.4.1
---------------------------

* Fixed inclusion path in pkg-config file

* Fixed extraneous epsilon path in hfst-lexc transducers

Noteworthy changes in release 2.4
---------------------------------

* Multiple compilation issued fixed w.r.t string.h inclusion inside hfst source
  tree without installed headers

* Added preliminary pkg-config support

* Optimizations for flag diacritics in optimized format; NB. the optimized
  format in 2.4 is not compatible with optimized format of previous releases

* More input and output formatting options for hfst-lookup

* New tool hfst-omor-evaluate for testing morphological transducers
  (an overblown glorified hfst-diff-test)

* Preliminary tool for killing evil flag diacritics

* Several minor bug fixes in composing intersection and flag diacritics

Noteworthy changes in release 2.3
---------------------------------

* Major fixes in flag diacritic handling for lookup and lookup-optimize

* New tool hfst-substitute for safe relabeling

* Bug fixes in hfst-compose-intersect w.r.t. any symbol and #

* Remove gettext support because of numerous portability issues

Noteworthy changes in release 2.2
---------------------------------

* Bug fixes in intersecting composition,  is_subset, negation, 

* Handling of flag diacritics included in lookup

* Remove readline because of numerous portability issues

* Remove alphabet and metadata from hfst-lexc

* hfst-lookup support for custom formatting of readings

* fixed hfst-twolc escaping in set constructions

Noteworthy changes in release 2.1
---------------------------------

* hfst-twolc and hfst-compose-intersect understand # like xerox originals

.. vim: set ft=rst:


hfst3/tools/src/hfst-regexp2fst.cc
//! @file hfst-regexp2fst.cc
//!
//! @brief regular expression compiling command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif


#include <iostream>
#include <fstream>

#include <vector>
#include <map>

using std::vector;
using std::pair;

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <getopt.h>
#include <math.h>
#include <errno.h>

#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "parsers/XreCompiler.h"
#include "hfst-commandline.h"
#include "hfst-program-options.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstOutputStream;
using hfst::HfstTokenizer;
using hfst::HfstTransducer;
using hfst::xre::XreCompiler;

static char *epsilonname=NULL;
static bool disjunct_expressions=false;
static bool line_separated = true;

//static unsigned int sum_of_weights=0;
static bool sum_weights=false;
static bool normalize_weights=false;
static bool logarithmic_weights=false;

static hfst::ImplementationType output_format = hfst::UNSPECIFIED_TYPE;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
        "Compile string pairs and pair-strings into transducer(s)\n"
        "\n", program_name); 
        print_common_program_options(message_out);
        print_common_unary_program_options(message_out); 
        fprintf(message_out, "String and format options:\n"
                "  -f, --format=FMT          Write result in FMT format\n"
                "  -j, --disjunct-strings    Disjunct all strings instead of "
                    "transforming each string into a separate transducer\n"
                "      --sum                 Sum weights of duplicate strings "
                    "instead of taking minimum\n"
                "      --norm                Divide each weight by sum "
                    "of all weights\n"
                "      --log                 Take negative logarithm "
                    "of each weight\n"
                "  -l, --line                Input is line separated\n"
                "  -S, --semicolon           Input is semicolon separated\n"
                "  -e, --epsilon=EPS         Map EPS as zero.\n");
        fprintf(message_out, "\n");

        fprintf(message_out, 
            "If OUTFILE or INFILE is missing or -, standard streams will be used.\n"
            "FMT must be name of a format usable by libhfst, such as "
            "openfst-tropical, sfst, foma or hfst-optimized-weighted\n"
            "If EPS is not defined, the default representation of @0@ is used\n"
            );

        fprintf(message_out, "Examples:\n"
            "  echo \"c:d a:o t:g\" | %s -l      create cat:dog fst\n"
            "  echo \"c:d a:o t:g ;\" | %s -S    same as pairstring\n"
            "\n", program_name, program_name);
        print_report_bugs();
        fprintf(message_out, "\n");
        print_more_info();
        fprintf(message_out, "\n");
}

int
parse_options(int argc, char** argv)
{
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
        {
        HFST_GETOPT_COMMON_LONG,
        HFST_GETOPT_UNARY_LONG,
          {"disjunct-strings", no_argument, 0, 'j'},
          {"epsilon", required_argument, 0, 'e'},
          {"sum", no_argument, 0, '1'},
          {"norm", no_argument, 0, '2'},
          {"log", no_argument, 0, '3'},
          {"line", no_argument, 0, 'l'},
          {"semicolon", no_argument, 0, 'S'},
          {"format", required_argument, 0, 'f'},
          {0,0,0,0}
        };
        int option_index = 0;
        char c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "je:123lSf:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }

        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case 'e':
            epsilonname = hfst_strdup(optarg);
            break;
        case '1':
            sum_weights = true;
            break;
        case '2':
            normalize_weights = true;
            break;
        case '3':
            logarithmic_weights = true;
            break;
        case 'j':
            disjunct_expressions = true;
            break;
        case 'S':
            line_separated = false;
            break;
        case 'l':
            line_separated = true;
            break;
        case 'f':
            output_format = hfst_parse_format_name(optarg);
            break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    if (output_format == hfst::UNSPECIFIED_TYPE)
      {
        verbose_printf("Output format not specified, "
             "defaulting to openfst tropical\n");
        output_format = hfst::TROPICAL_OPENFST_TYPE;
      }
    return EXIT_CONTINUE;
}

int
process_stream(HfstOutputStream& outstream)
{
  size_t transducer_n = 0;
  char* line = 0;
  size_t len = 0;
  unsigned int line_count = 0;
  XreCompiler comp(output_format);
  HfstTransducer disjunction(output_format);
  //outstream.open();
  int delim = '\n';
  if (line_separated)
    {
      delim = '\n';
    }
  else
    {
      delim = ';';
    }
  while (hfst_getdelim(&line, &len, delim, inputfile) != -1)
    {
      transducer_n++;
      line_count++;
      HfstTransducer* compiled;
      verbose_printf("Compiling expression %u\n", line_count);
      compiled = comp.compile(line);
      if (disjunct_expressions)
        {
          disjunction.disjunct(*compiled);
        }
      else
        {
          char* name = static_cast<char*>(malloc(sizeof(char)*(strlen("hfst-regexp2fst ") + strlen(line) + 1)));
          if (sprintf(name, "hfst-regexp2fst %s", line) > 0)
            {
              compiled->set_name(name);
            }
          else
            {
              compiled->set_name("hfst-regexp2fst <error in sprintf>");
            }
          outstream << *compiled;
        }
      delete compiled;
    }
  if (disjunct_expressions)
    {
      char* name = static_cast<char*>(malloc(sizeof(char)*(strlen("hfst-regexp2fst ") + strlen(inputfilename) + 1)));
      if (sprintf(name, "hfst-regexp2fst %s", inputfilename) > 0)
        {
          disjunction.set_name(name);
        }
      else
        {
          disjunction.set_name("hfst-regexp2fst <error in sprintf>");
        }
      outstream << disjunction;
    }
  free(line);
  return EXIT_SUCCESS;
}


int main( int argc, char **argv ) 
{
  hfst_set_program_name(argv[0], "0.1", "Regexp2Fst");
  int retval = parse_options(argc, argv);
  if (retval != EXIT_CONTINUE)
    {
      return retval;
    }
  // close buffers, we use streams
  if (outfile != stdout)
    {
      fclose(outfile);
    }
  verbose_printf("Reading from %s, writing to %s\n", 
                 inputfilename, outfilename);
  // here starts the buffer handling part
  HfstOutputStream* outstream = (outfile != stdout) ?
        new HfstOutputStream(outfilename, output_format) :
        new HfstOutputStream(output_format);
  process_stream(*outstream);
  free(inputfilename);
  free(outfilename);
  return EXIT_SUCCESS;
}

